client/beefy/src/import.rs

```diff
@@ -0,0 +1,197 @@
+{
+	type Error = ConsensusError;
+	type Transaction = TransactionFor<Client, Block>;
+
+	async fn import_block(
+		&mut self,
+		block: BlockImportParams<Block, Self::Transaction>,
+		new_cache: HashMap<well_known_cache_keys::Id, Vec<u8>>,
+	) -> Result<ImportResult, Self::Error> {
+		let hash = block.post_hash();
+		let number = *block.header.number();
+		let justifications = block.justifications.clone();
+		// If this block contains a beefy authority set change then it must have a beefy
+		// justification
+		let change = find_beefy_authority_set_change::<Block>(&block.header);
+		// Run inner block import
+		let import_result = self.inner.import_block(block, new_cache).await?;
+		// Try importing beefy justification
+		let beefy_justification =
+			justifications.and_then(|just| just.into_justification(BEEFY_ENGINE_ID));
+
+		if change.is_some() && beefy_justification.is_none() {
+			return Err(ConsensusError::ClientImport(
+				"Missing beefy justification in authority set change block".to_string(),
+			))
+		}
+		if let Some(beefy_justification) = beefy_justification {
+			self.import_justification(hash, number, (BEEFY_ENGINE_ID, beefy_justification))?;
+		}
+		Ok(import_result)
+	}
+
+	async fn check_block(
+		&mut self,
+		block: BlockCheckParams<Block>,
+	) -> Result<ImportResult, Self::Error> {
+		self.inner.check_block(block.clone()).await
+	}
+}
+
+#[async_trait::async_trait]
+impl<BE, Block: BlockT, Client, I> JustificationImport<Block>
+	for BeefyBlockImport<BE, Block, Client, I>
+where
+	BE: Backend<Block>,
+	Client: BeefyClient<Block, BE> + ProvideRuntimeApi<Block>,
+	Client::Api: BeefyApi<Block>,
+	I: JustificationImport<Block, Error = ConsensusError> + Send + Sync,
+{
+	type Error = ConsensusError;
+
+	async fn on_start(&mut self) -> Vec<(Block::Hash, NumberFor<Block>)> {
+		self.inner.on_start().await
+	}
+
+	async fn import_justification(
+		&mut self,
+		hash: Block::Hash,
+		number: NumberFor<Block>,
+		justification: Justification,
+	) -> Result<(), Self::Error> {
+		// Try Importing Beefy justification
+		BeefyBlockImport::import_justification(self, hash, number, justification.clone())?;
+		// Importing for inner BlockImport
+		self.inner.import_justification(hash, number, justification).await
+	}
+}
+
+impl<BE, Block: BlockT, Client, I> BeefyBlockImport<BE, Block, Client, I>
+where
+	BE: Backend<Block>,
+	Client: BeefyClient<Block, BE> + ProvideRuntimeApi<Block>,
+	Client::Api: BeefyApi<Block>,
+{
+	/// Import a block justification.
+	fn import_justification(
+		&mut self,
+		hash: Block::Hash,
+		number: NumberFor<Block>,
+		justification: Justification,
+	) -> Result<(), ConsensusError> {
+		if justification.0 != BEEFY_ENGINE_ID {
+			return Ok(())
+		}
+
+		// This function assumes the Block should have already been imported
+		let at = BlockId::hash(hash);
+		let validator_set = self
+			.client
+			.runtime_api()
+			.validator_set(&at)
+			.map_err(|e| ConsensusError::ClientImport(e.to_string()))?;
```

Comment:
We should be verifying that the previous authority set has signed off on the hand-off instead.